README.md
# Wait for Netlify Deploy — A GitHub Action ⏱

Do you have other Github actions (Lighthouse, Cypress, etc) that depend on the Netlify Preview URL? This action will wait until the preview URL is available before running the next task.

This is a fork of [JosephDuffy/wait-for-netlify-action](https://github.com/JosephDuffy/wait-for-netlify-action) that uses the deployment for the commit, rather than for the PR. This fork fixes a couple issues retrieving the GitHub SHA being built, the deployment commit URL, and `max_timeout` setting.

This action uses the Netlify API to always retrieve the correct deployment being built. You will need to generate a [Personal Access Token](https://app.netlify.com/user/applications/personal) to use and pass it as the `NETLIFY_TOKEN` environment variable.

## Env

### `NETLIFY_TOKEN`

**Required.** Your Netlify [Personal Access Token](https://app.netlify.com/user/applications/personal) to use for API access. This should be set as a GitHub secret, see example.

## Inputs

### `site_name`

**Required** The name of the Netlify site to reach `https://{site_name}.netlify.com`

### `max_timeout`

Optional — The amount of time to spend waiting on Netlify. Defaults to `60` seconds

## Outputs

### `url`

The Netlify deploy preview url that was deployed.

### `site_id`

The Netlify site ID that was deployed.

### `deploy_id`

The Netlify deployment ID that was deployed.

## Example usage

Basic Usage

```yaml
steps:
  - name: Waiting for Netlify Preview
    uses: kamranayub/wait-for-netlify-action@v2.1.0
    id: wait-for-netflify-preview
    with:
      site_name: "YOUR_SITE_NAME"
      max_timeout: 60
    env:
      NETLIFY_TOKEN: ${{secrets.NETLIFY_TOKEN}}
```

<details>
<summary>Complete example with Lighthouse</summary>
<br />

```yaml
name: Lighthouse

on: [pull_request]

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v1
      - name: Use Node.js 12.x
        uses: actions/setup-node@v1
        with:
          node-version: 12.x
      - name: Install
        run: |
          npm ci
      - name: Build
        run: |
          npm run build
      - name: Waiting for 200 from the Netlify Preview
        uses: kamranayub/wait-for-netlify-action@v2.1.0
        id: wait-for-netflify-preview
        with:
          site_name: "YOUR_SITE_NAME"
        env:
          NETLIFY_TOKEN: ${{secrets.NETLIFY_TOKEN}}
      - name: Lighthouse CI
        run: |
          npm install -g @lhci/cli@0.3.x
          lhci autorun --upload.target=temporary-public-storage --collect.url=${{ steps.wait-for-netflify-preview.outputs.url }} || echo "LHCI failed!"
        env:
          LHCI_GITHUB_APP_TOKEN: ${{ secrets.LHCI_GITHUB_APP_TOKEN }}
```

</details>


CHANGELOG.md
# [2.1.0](https://github.com/kamranayub/wait-for-netlify-action/compare/v2.0.3...v2.1.0) (2022-10-21)


### Features

* query *all* sites and match on exact site name ([#7](https://github.com/kamranayub/wait-for-netlify-action/issues/7)) ([40462bd](https://github.com/kamranayub/wait-for-netlify-action/commit/40462bd43a592d2926663e1e539f3751f768a1d9))

## [2.0.3](https://github.com/kamranayub/wait-for-netlify-action/compare/v2.0.2...v2.0.3) (2020-07-31)


### Bug Fixes

* should work with master or production branches now ([324e205](https://github.com/kamranayub/wait-for-netlify-action/commit/324e205bbe6802718a113dccf55ef3a556b967e7))
* **package:** typos and path ([8d57496](https://github.com/kamranayub/wait-for-netlify-action/commit/8d57496916c8b82099bff81062f1fe4f1ee160d1))
* should work on non-pr builds ([8f01806](https://github.com/kamranayub/wait-for-netlify-action/commit/8f018069ef245071a8a91c7722365020c9576ec8))

## [2.0.2](https://github.com/kamranayub/wait-for-netlify-action/compare/v2.0.1...v2.0.2) (2020-07-31)


### Bug Fixes

* **package:** typos and path ([4d21387](https://github.com/kamranayub/wait-for-netlify-action/commit/4d213878739b2b970f2d67b128a332411aa508b3))

## [2.0.1](https://github.com/kamranayub/wait-for-netlify-action/compare/v2.0.0...v2.0.1) (2020-07-31)


### Bug Fixes

* **package:** always build on pre-commit ([1fec6dd](https://github.com/kamranayub/wait-for-netlify-action/commit/1fec6dd7db4396975eb2929e7a6777bcf67288a9))

# [2.0.0](https://github.com/kamranayub/wait-for-netlify-action/compare/v1.0.0...v2.0.0) (2020-07-31)


### Bug Fixes

* **package:** increment major ([74b5c7c](https://github.com/kamranayub/wait-for-netlify-action/commit/74b5c7c05fcce2bc318bb95067948fb4797c8b6b))


### BREAKING CHANGES

* **package:** force update for tagging purposes

# 1.0.0 (2020-07-31)


### Bug Fixes

* async/await all the things ([17c79ad](https://github.com/kamranayub/wait-for-netlify-action/commit/17c79ad645470ec3b7e3227a8bb6ce7fc005e14d))
* check-in node_modules ([0f7e84d](https://github.com/kamranayub/wait-for-netlify-action/commit/0f7e84d74c9d4346904ef37138426a7f5d8ec88c))
* commit node_modules ([6ba8bb0](https://github.com/kamranayub/wait-for-netlify-action/commit/6ba8bb03164465254043207015218519359ad088))
* max_timeout, github sha ([4f68c63](https://github.com/kamranayub/wait-for-netlify-action/commit/4f68c63fc14e989b504b6b3bc4635f276b3767d0))
* move to async function ([4fb77dd](https://github.com/kamranayub/wait-for-netlify-action/commit/4fb77dd38734cf7567c10ec9ed3768ab31e2d148))
* params to axios.get ([cd0bdf8](https://github.com/kamranayub/wait-for-netlify-action/commit/cd0bdf80eb7de68410cc00681781c733f2d8206a))
* pass netlify token to auth header ([9c03295](https://github.com/kamranayub/wait-for-netlify-action/commit/9c03295155a77476b56550255cbeafabba3d222a))
* updating the "uses" path ([f16b004](https://github.com/kamranayub/wait-for-netlify-action/commit/f16b004f304dd3dbe6a83ff81f5fe47a91a389f4))


### Features

* adding complete example ([7bee90c](https://github.com/kamranayub/wait-for-netlify-action/commit/7bee90c1b2cdf840ea7b59845992af5be171e6e5))
* adding error handling ([2fbf1a3](https://github.com/kamranayub/wait-for-netlify-action/commit/2fbf1a3050c92c8e6c5e15a3fc0459ebdeeaa9fe))
* adding example action ([9265032](https://github.com/kamranayub/wait-for-netlify-action/commit/9265032bccf41044d1992e514959e7156670e001))
* adding output. updating action yaml ([c09179c](https://github.com/kamranayub/wait-for-netlify-action/commit/c09179c52e7d2a9429da180b9b38ecf06db47065))
* basic logic to wait for a 200 response ([771ad34](https://github.com/kamranayub/wait-for-netlify-action/commit/771ad349cac19486730ee3f3337ca6640ddcb30a))
* configure hello world action ([dc1ad01](https://github.com/kamranayub/wait-for-netlify-action/commit/dc1ad01b9085dd61303746fd6c2eb42d660fb578))
* configure local workflow ([#1](https://github.com/kamranayub/wait-for-netlify-action/issues/1)) ([2d43cf2](https://github.com/kamranayub/wait-for-netlify-action/commit/2d43cf2d348037c32c3c336f216959c9b866ef4e))
* should timeout for 404 url ([1ed1b71](https://github.com/kamranayub/wait-for-netlify-action/commit/1ed1b711aabb3d8a2c933175f29f47a45a2b7ed1))
* use Netlify API ([80aff93](https://github.com/kamranayub/wait-for-netlify-action/commit/80aff936c285354a0dadc2ea4326141bc01474e1))


index.js
/* eslint-disable no-console, no-await-in-loop  */
const core = require('@actions/core');
const github = require('@actions/github');
const axios = require('axios');

function getNetlifyUrl(url) {
  return axios.get(url, {
    headers: {
      Authorization: `Bearer ${process.env.NETLIFY_TOKEN}`,
    },
  });
}

const waitForUrl = async (url, MAX_TIMEOUT) => {
  const iterations = MAX_TIMEOUT / 3;
  for (let i = 0; i < iterations; i += 1) {
    try {
      await axios.get(url);
      console.log('Url fetch completed, done!');
      return;
    } catch (e) {
      console.log('Url unavailable, retrying...');
      await new Promise((r) => setTimeout(r, 3000));
    }
  }
  core.setFailed(`Timeout reached: Unable to connect to ${url}`);
};

const run = async () => {
  try {
    const netlifyToken = process.env.NETLIFY_TOKEN;
    // In a PR, github.contex.sha refers to the last merge commit SHA
    // not the *latest commit* of the PR which is what Netlify uses. Instead,
    // have to use github.context.payload.pull_request.head.sha
    // See: https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads#pull_request
    const isPr = 'pull_request' in github.context.payload;
    const commitSha = isPr ? github.context.payload.pull_request.head.sha : github.context.sha;
    const MAX_TIMEOUT = Number(core.getInput('max_timeout')) || 60;
    const siteName = core.getInput('site_name');
    if (!netlifyToken) {
      core.setFailed('Please set NETLIFY_TOKEN env variable to your Netlify Personal Access Token secret');
    }

    if (!commitSha) {
      core.setFailed('Could not determine GitHub commit');
    } else {
      console.log('Using SHA', commitSha, isPr ? 'from PR' : '');
    }
    if (!siteName) {
      core.setFailed('Required field `site_name` was not provided');
    }

    const { data: netlifySites } = await getNetlifyUrl(
      `https://api.netlify.com/api/v1/sites?filter=all&name=${siteName}`,
    );
    if (!netlifySites || netlifySites.length === 0) {
      core.setFailed(`Could not find Netlify site with the name ${siteName}`);
    }
    const { site_id: siteId } = netlifySites.find((site) => site.name === siteName);
    core.setOutput('site_id', siteId);

    const { data: netlifyDeployments } = await getNetlifyUrl(`https://api.netlify.com/api/v1/sites/${siteId}/deploys`);

    if (!netlifyDeployments) {
      core.setFailed('Failed to get deployments for site');
    }

    // Most likely, it's the first entry in the response
    // but we correlate it just to be safe
    const commitDeployment = netlifyDeployments.find(
      (d) => d.commit_ref === commitSha && d.context === (isPr ? 'deploy-preview' : 'production'),
    );
    if (!commitDeployment) {
      core.setFailed(`Could not find deployment for commit ${commitSha}`);
    }
    core.setOutput('deploy_id', commitDeployment.id);

    // At this point, we have enough info to where
    // we could wait for the deployment state === "ready"
    // but it's probably more reliable to wait for the URL
    // to be available.
    //
    // This could be enhanced to wait for the deployment status
    // and then wait once again for the URL to return 200.
    const url = `https://${commitDeployment.id}--${siteName}.netlify.app`;
    core.setOutput('url', url);
    console.log(`Waiting for a 200 from: ${url}`);
    await waitForUrl(url, MAX_TIMEOUT);
  } catch (error) {
    core.setFailed(error.message);
  }
};

run();
